onmt/train_single.py

```diff
@@ -1,17 +1,44 @@
 #!/usr/bin/env python
 """Training on a single process."""
+import sys
 import torch
-
+from onmt.constants import CorpusTask
 from onmt.inputters.inputter import IterOnDevice
+from onmt.inputters import build_vocab
 from onmt.model_builder import build_model
 from onmt.utils.optimizers import Optimizer
 from onmt.utils.misc import set_random_seed
 from onmt.trainer import build_trainer
 from onmt.models import build_model_saver
 from onmt.utils.logging import init_logger, logger
 from onmt.utils.parse import ArgumentParser
-
 from onmt.inputters.dynamic_iterator import build_dynamic_dataset_iter
+from onmt.transforms import make_transforms, save_transforms, \
+    get_transforms_cls
+from onmt.inputters.corpus import save_transformed_sample
+
+
+def prepare_transforms(opt, vocabs):
+    """Prepare or dump fields & transforms before training."""
+    transforms_cls = get_transforms_cls(opt._all_transform)
+    # specials = get_specials(opt, transforms_cls)
```

Comment:
Yes I missed that, but honestly I don't know exactly where this is used. Can you give me more info ?

Reply:
@Zenglinxiao is there a test somewhere that I could look at to test if and when thoss specials depending on transforms may play a role and be not empty ? thanks

Reply:
@vince62s Hi Vincent, you might want to refer to this: https://github.com/OpenNMT/OpenNMT-py/blob/990dcf6edf8c2f30328d03d2c54c67184df0aec1/onmt/tests/test_transform.py#L42-L56

This `get_specials` method basically automatically adds special tokens newly created by the transform function itself. For example, if we randomly mask source sentences, then it will return `MASK` token to the source side to be merged into the source dictionary.